src/storage-operators/src/s3_oneshot_sink.rs

```diff
@@ -56,33 +61,301 @@ pub fn copy_to<G, F>(
     G: Scope<Timestamp = Timestamp>,
     F: FnOnce(Result<u64, String>) -> () + 'static,
 {
-    let mut scope = input_collection.scope();
+    let scope = input_collection.scope();
+
+    let s3_key_manager = S3KeyManager::new(&sink_id, &connection_details.uri);
+
+    let start_stream = render_initialization_operator(
+        scope.clone(),
+        connection_context.clone(),
+        aws_connection.clone(),
+        connection_id,
+        sink_id,
+        s3_key_manager.clone(),
+    );
+
+    let completion_stream = render_upload_operator(
+        scope.clone(),
+        connection_context.clone(),
+        aws_connection.clone(),
+        connection_id,
+        connection_details,
+        sink_id,
+        input_collection,
+        err_collection,
+        up_to,
+        start_stream,
+    );
+
+    render_completion_operator(
+        scope,
+        connection_context,
+        aws_connection,
+        connection_id,
+        sink_id,
+        s3_key_manager,
+        completion_stream,
+        onetime_callback,
+    );
+}
+
+/// Renders the 'initialization' operator, which does work on the leader worker only.
+/// The leader worker checks the S3 path for the sink to ensure it's empty
+/// (aside from files written by other instances of this sink), and writes an INCOMPLETE
+/// sentinel file to indicate to the user that the upload is in-progress.
+///
+/// The INCOMPLETE sentinel is used to provide a single atomic operation that a user
+/// can wire up a notification on, to know when it is safe to start ingesting the
+/// data written by this sink to S3. Since the DeleteObject of the INCOMPLETE sentinel
+/// will only trigger one S3 notification, even if it's performed by multiple replicas
+/// it simplifies the user ergonomics by only having to listen for a single event
+/// (a PutObject sentinel would trigger once for each replica).
+///
+/// Returns a stream with a result object indicating the success or failure of the
+/// initialization operation.
+fn render_initialization_operator<G>(
+    scope: G,
+    connection_context: ConnectionContext,
+    aws_connection: AwsConnection,
+    connection_id: GlobalId,
+    sink_id: GlobalId,
+    s3_key_manager: S3KeyManager,
+) -> Stream<G, Result<(), String>>
+where
+    G: Scope<Timestamp = Timestamp>,
+{
+    let worker_id = scope.index();
+    let num_workers = scope.peers();
+    let leader_id = usize::cast_from((sink_id, "initialization").hashed()) % num_workers;
+    let is_leader = worker_id == leader_id;
+
+    let mut builder =
+        AsyncOperatorBuilder::new("CopyToS3-initialization".to_string(), scope.clone());
+
+    let (mut start_handle, start_stream) = builder.new_output();
+
+    builder.build(move |caps| async move {
+        let [start_cap] = caps.try_into().unwrap();
+
+        // fallible async block to use the `?` operator for convenience
+        let leader_work = || async move {
```

Comment:
I think you don't need the closure here

```suggestion
        let leader_work = async move {
```